predicthq/endpoints/decorators.py

```diff
@@ -46,82 +48,42 @@ def _process_kwargs(kwargs, separator="__"):
-                    endpoint.mutate_bool_to_default_for_type(data)
-                model = schema()
-                model.import_data(data, strict=True, partial=False)
-                model.validate()
-            except SchematicsDataError as e:
-                raise ValidationError(e.messages)
+            data = _process_kwargs(kwargs)
+            if hasattr(endpoint, "mutate_bool_to_default_for_type"):
+                endpoint.mutate_bool_to_default_for_type(data)
 
             if query_string:
-                params = _to_url_params(model.to_primitive(role=role))
-            else:
-                params = model.to_primitive(role=role)
+                data = _to_url_params(data=data)
 
-            return f(endpoint, *args, **params)
+            return f(endpoint, *args, **data)
 
         return wrapper
 
     return decorator
 
 
-def returns(schema_class):
+def returns(model_class):
     def decorator(f):
         @functools.wraps(f)
         def wrapper(endpoint, *args, **kwargs):
 
-            schema = getattr(endpoint.Meta, f.__name__, {}).get("returns", schema_class)
+            model = getattr(endpoint.Meta, f.__name__, {}).get("returns", model_class)
 
             data = f(endpoint, *args, **kwargs)
             try:
-                model = schema()
-                model._endpoint = endpoint
-
-                # if schema class is a ResultSet, tell it how to load more results
-                if issubclass(schema_class, ResultSet):
-                    model._more = functools.partial(wrapper, endpoint)
-
-                    # if results are of type Model, make sure to set the endpoint on each item
-                    if (
-                        data is not None
-                        and "results" in data
-                        and hasattr(model._fields["results"], "model_class")
-                        and issubclass(model._fields["results"].model_class, Model)
-                    ):
-
-                        def initialize_result_type(item_data):
-                            item = model._fields["results"].model_class(item_data, strict=False)
-                            item._endpoint = endpoint
-                            return item
-
-                        # Use generator so results are not iterated over more than necessary
-                        data["results"] = (initialize_result_type(item_data) for item_data in data["results"])
-
-                model.import_data(data, strict=False)
-                model.validate()
-            except SchematicsDataError as e:
-                raise ValidationError(e.messages)
-            return model
+                loaded_model = model(**data) if data else model()
+                loaded_model._more = functools.partial(wrapper, endpoint)
+                loaded_model._endpoint = endpoint
+                return loaded_model
+            except PydanticValidationError as e:
+                if not data:
+                    return
```

Comment:
Good point raising this one. The underlying logic was not correct, so I updated it to re-raise any validation error.

The current behaviour should be as follows:
- If the requests issued an HTTP error, then either raise a `ClientError` or `ServerError` depending on the error code
- Else, data should be expected. Either empty or full data, so deserialise it into Pydantic models (which should accept an empty dataset if it can happen). This will raise a `ValidationError` error if the data returned by the API is not complying the SDK Pydantic models restrictions (which we ideally never want to happen - this will be fine tuned through testing phase)

Reply:
Sounds good @lpouvreau Thanks for the detailed explanation! :star: